File: summary/rust/17_shakutori.rs
fn shakutori(arr: &[i32], k: i32) -> i32 {
    let n = arr.len();
    let mut res = 0;
    let mut right = 0;
    let mut sum = 0;

    for left in 0..n {
        while right < n && sum + arr[right] <= k {
            sum += arr[right];
            right += 1;
        }

        res += (right - left) as i32;

        if right == left {
            right += 1;
        } else {
            sum -= arr[left];
        }
    }

    res
}

fn main() {
    let input_array: Vec<i32> = vec![1, 2, 3, 4, 5];
    let k = 5;
    let result = shakutori(&input_array, k);
    println!("Number of subarrays with sum <= {}: {}", k, result);
}

File: summary/elixir/17_shakutori.exs
defmodule Shakutori do
  def solve(arr, k) do
    n = length(arr)
    do_shakutori(arr, k, n, 0, 0, 0, 0)
  end

  defp do_shakutori(arr, k, n, left, right, sum, res) when left < n do
    {new_right, new_sum} = advance_right(arr, k, n, right, sum)
    new_res = res + (new_right - left)

    if new_right == left do
      do_shakutori(arr, k, n, left + 1, new_right + 1, 0, new_res)
    else
      do_shakutori(arr, k, n, left + 1, new_right, new_sum - Enum.at(arr, left), new_res)
    end
  end

  defp do_shakutori(_, _, _, _, _, _, res), do: res

  defp advance_right(arr, k, n, right, sum) when right < n and sum + Enum.at(arr, right) <= k do
    advance_right(arr, k, n, right + 1, sum + Enum.at(arr, right))
  end
  defp advance_right(_, _, _, right, sum), do: {right, sum}
end

# Example usage:
input_array = IO.gets("Enter the array elements separated by space: ")
              |> String.trim()
              |> String.split()
              |> Enum.map(&String.to_integer/1)

k = IO.gets("Enter the value of k: ")
    |> String.trim()
    |> String.to_integer()

result = Shakutori.solve(input_array, k)
IO.puts("Number of subarrays with sum <= #{k}: #{result}")
